### ISSUE TITLE
500 server error when attempt to create account (API)

### ISSUE THREAD

**Issue by kdelee** [cloudigrade/cloudigrade#286]
## Summary

When I attempt to create an account with a `POST` to `api/v1/account/`, I get a 500 internal server error instead of some more informative error that tells me what I did wrong.

## Steps to Reproduce

Send a `POST` to `/api/v1/account/` using proper authentication with data:

```
'{"resourcetype": "AwsAccount", "account_arn": "arn:aws:iam::518028203513:role/grant_cloudi_to_372779871274"}'
```

(I got this arn from the api example docs)

## Expected Result

- Either 200 success and a new account or a 400 with an error message that tells me what I did wrong.

## Actual Result

```
response code : 500

text_error_message : '<h1>Server Error (500)</h1>'
```

## Additional context

Serving locally in openshift, using 5428f7cbffde008698bb0de8153f1123d237ec2c .
Was attempting to create accounts with API in order to test ability to list them.

This is blocking me from testing listing the accounts because I would like to do the following:

1) Create X number of accounts
2) Request list of accountds
3) Validate all the accounts are in the list

**Comment by kdelee** [comment]
@infinitewarp another thing is that a `POST` to `api/v1/account/` is only told one item that is missing from the request. In my experience it is possible and preferable to return a list of all missing items. So right now if I send an empty body with a `POST` I only get back the message: `{'resourcetype': 'This field is required'}` but I should get back `[{'resourcetype': 'This field is required'}, {'account_arn': ['This field is required.']} ]` or something like that.

Do you want to roll that into this as it is a part of the `400` response I should (perhaps) be getting if I'm doing something wrong, or do you want a separate issue?

Side note: if endpoints could report the fields they accept when you send `OPTIONS` to the endpoint that would be a great way to make the API more discoverable for QE and UI devs. i.e. right now I'm guessing that `resourcetype` has a finite list of valid values (`['AwsAccount']`)

**Comment by kdelee** [comment]
@infinitewarp 

Recreated again this morning with fresh environment. My set up was this:

1) Pull lastest master branch
2) make oc-up-all
3) make oc-user
4) make oc-user-authenticate & save token
5) `http post http://$CLOUDIGRADE_BASE_URL/api/v1/account/ Authorization:"Token ${CLOUDIGRADE_TOKEN}" resourcetype="AwsAccount" account_arn="arn:aws:iam::518028203513:role/grant_cloudi_to_372779871274"`

**Comment by infinitewarp** [comment]
I destroyed my OpenShift environment (`make oc-clean`) so it would rebuild and redeploy everything from scratch, and I followed your commands exactly with `cloudigrade-myproject.127.0.0.1.nip.io` for `$CLOUDIGRADE_BASE_URL` and the created token for `${CLOUDIGRADE_TOKEN}`. Yet, I'm still not able to reproduce this. 😕

I'm thinking there must be something wrong in or missing from your environment. I'll draw up some notes for things to look for in your environment and maybe add some logging to try to identify where things are breaking.

Regarding your previous comment about listing missing things in the 400 error, yes, please track that in a separate issue if you think it's a bug. I believe what you ask for is actually the correct default behavior with DRF validators, *but* keep in mind that there are layers to validation in this case, and the polymorphic resource must first be correctly identified (using a valid `resourcetype`) *before* we can tell the client what else that specific resource type needs (e.g. `account_arn`).

**Comment by infinitewarp** [comment]
Just to verify that you have valid `AWS_` environment variables set locally, what do you get when you run these commands from your local terminal, @kdelee ? To ensure that the environment variables are the same, please execute these from the same shell environment where you would run your `make` commands to start up and deploy to your openshift cluster.

```
aws iam list-account-aliases
aws iam get-user
```

Maybe redact parts of your user info from the output if you like. 🙂

**Comment by infinitewarp** [comment]
Also, given the ARN in your example commands, does this command work for you @kdelee ?

```
aws sts assume-role --role-arn "arn:aws:iam::518028203513:role/grant_cloudi_to_372779871274" --role-session-name test1
```

Again, you may want to redact parts of the output if you share it here. 🙂

**Comment by infinitewarp** [comment]
I think I may have found a way to reproduce this (or at least something remarkably close).

If I try to register a cloud account with a well-structured but bogus (it doesn't actually exist) ARN, then I also get an HTTP 500 error. I'll wrap that up with appropriate handling, and hopefully that'll fix the original issue here too.

**Comment by kdelee** [comment]
On Thu, May 17, 2018, 10:34 AM Brad Smith <notifications@github.com> wrote:

> I think I may have found a way to reproduce this (or at least something
> remarkably close).
>
> If I try to register a cloud account with a well-structured but bogus (it
> doesn't actually exist) ARN, then I also get an HTTP 500 error. I'll wrap
> that up with appropriate handling, and hopefully that'll fix the original
> issue here too.
>

Cool. I would follow up with @elyezer because this will relate to his work
on intregrade issues.

One thing that would be good to clarify for testing is if the arn  is
validated with aws  at creation time. Sounds like you are saying it is.

**Comment by infinitewarp** [comment]
Demo of the fixes and changes around exception handling: https://asciinema.org/a/182337

Yes, the ARN is fully validated for permissions when you register the cloud account for it. Previously, we assumed we at least had access to the ARN and went straight on to checking the actions on the Policy that we expect to need. One of the changes I made with my fix is to verify that we have access _before_ checking the Policy actions, and raise an appropriate error if we don't have that access.

"Unfortunately", I don't have a demo of a _real_ 500 error in there because I couldn't think of a way to actually _cause_ a 500 error now that I've cleaned up all the issues around the initial ARM permissions checking. 😅
